### PR TITLE
Switch the ADAL android to Android Studio

### PR DESCRIPTION
**Pull request by heidijinxujia**
The goal for this PR is to configure the gradle build for the android studio. Any code change should be in their own sub-branch from branch android-studio.

• this branch is a sub-branch from dev
• configure Findbugs, PMD, Android Lint and Checkstyle into the gradle build
• create the jar and aar files
• unify the names of the output
• remove Travis

### CODE REVIEW

File: adal/proguard-rules.pro [hunk 1]
```diff
@@ -1,11 +1,8 @@
-# To enable ProGuard in your project, edit project.properties
-# to define the proguard.config property as described in that file.
-#
 # Add project specific ProGuard rules here.
```
Comment (weijjia) [inline adal/proguard-rules.pro]: Do we have any proguard rules here? I wonder if we don't have rules, do we actually need it?

**Comment by weijjia** [comment]
If we're going to merge the android studio config into dev now, just want to do sanity check and please make sure we have the correct setting for publish to maven central, then the next release for 1.1.17 will be pushed out with gradle build.

**Comment by weijjia** [comment]
Also, IMO if pushing the change to dev, we should have travis configured. This just to make sure we won't have any PRs break our build.

File: adal/build.gradle [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+group = 'com.microsoft.aad'
+version = '1.1.16'
+
+buildscript {
+    repositories {
+        jcenter()
+    }
+    dependencies {
+            classpath 'com.github.dcendents:android-maven-gradle-plugin:1.3'
```
Comment (weijjia) [inline adal/build.gradle]: nit: use space instead of tab

File: adal/build.gradle [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    defaultConfig {
+        minSdkVersion 18
+        targetSdkVersion 21
+        versionCode 1
+        versionName "1.1.16"
+        project.archivesBaseName = "adal"
+        project.version = android.defaultConfig.versionName
+    }
+    buildTypes {
```
Comment (weijjia) [inline adal/build.gradle]: nit: new line

File: adal/build.gradle [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+        }
+    }
+    sourceSets {
```
Comment (weijjia) [inline adal/build.gradle]: nit: new line

File: adal/build.gradle [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    libraryVariants.all { variant ->
+        variant.outputs.each { output ->
+            def outputFile = output.outputFile
+            if (outputFile != null && outputFile.name.endsWith('.aar')) {
+                def fileName = "${archivesBaseName}-${version}.aar"
+                output.outputFile = new File(outputFile.parent, fileName)
+            }
+        }
+    }
+
```
Comment (weijjia) [inline adal/build.gradle]: nit: extra line

File: adal/build.gradle [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+dependencies {
+    compile fileTree(dir: 'libs', include: ['*.jar'])
```
Comment (weijjia) [inline adal/build.gradle]: why do we need to compile the jars in lib folder?

File: adal/build.gradle [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    androidTestCompile 'org.mockito:mockito-core:1.10.19'
+    androidTestCompile 'com.google.dexmaker:dexmaker:1.2'
+    androidTestCompile 'com.google.dexmaker:dexmaker-mockito:1.2'
+    compile 'com.android.support:appcompat-v7:21.0.3'
+    compile 'com.google.code.gson:gson:2.2.4'
+    testCompile 'junit:junit:4.12'
```
Comment (weijjia) [inline adal/build.gradle]: nit: but I prefer we separate the dependency like this:
// dependencies required by tests
list of the test dependencies

// dependencies required for compile
list of the dependencies required for compile

File: adal/build.gradle [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    destinationDir = reporting.file("$project.buildDir/outputs/jar/")
+}
+
+task javadoc(type: Javadoc) {
+    source = android.sourceSets.main.java.srcDirs
+    classpath = configurations.compile
+    destinationDir = reporting.file("$project.buildDir/outputs/jar/")
+}
+
+task javadocJar(type: Jar, dependsOn: javadoc) {
```
Comment (weijjia) [inline adal/build.gradle]: what will be the output name for sourcejar and javadoc jar?

File: adal/proguard-rules.pro (hunk 1 above)
Comment (heidijinxujia) [inline adal/proguard-rules.pro]: "The ProGuard tool shrinks, optimizes, and obfuscates your code by removing unused code and renaming classes, fields, and methods with semantically obscure names." I don't think we have the  need to use it now. I remove it.

**Comment by heidijinxujia** [comment]
OK. So there are two remaining things need to be done here. 
1. Deploying to maven central with Gradle
2. Travis configuration. 
@weijjia @RandalliLama

File: adal/proguard-rules.pro (hunk 1 above)
Comment (weijjia) [inline adal/proguard-rules.pro]: I know what proguards does, It's fine and nice to use it, most apps turn on proguard for release build. But we don't have any rules in the proguard file.

File: sample/src/test/java/com/microsoft/aad/adal/sample/ExampleUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package com.microsoft.aad.adal.sample;
+
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+/**
+ * To work on unit tests, switch the Test Artifact in the Build Variants view.
+ */
+public class ExampleUnitTest {
```
Comment (RandalliLama) [inline sample/src/test/java/com/microsoft/aad/adal/sample/ExampleUnitTest.java]: Does this file need to be in the repo or did some tool generate it.  It doesn't seem like we need a sample unit test.

File: adal/build.gradle (hunk 2 above)
Comment (RandalliLama) [inline adal/build.gradle]: @weijjia @heidijinxujia  That's not a nit.  That's a rule.

File: adal/build.gradle (hunk 6 above)
Comment (RandalliLama) [inline adal/build.gradle]: @weijjia I'm not arguing with you, you are probably right, but could you suggest what you think would be a better place?

File: adal/build.gradle (hunk 7 above)
Comment (RandalliLama) [inline adal/build.gradle]: agree

**Comment by RandalliLama** [comment]
@heidijinxujia Don't change the flag yet.  We can change that back as soon as we fix all the issues.  Which, of course, we will do before the next release.  I don't think we need to do it all in one PR though.  So don't bother with your #3 yet.

File: adal/build.gradle (hunk 6 above)
Comment (weijjia) [inline adal/build.gradle]: Well, I simply don't know why we need to lib folder. Do we have any particular jar to compile that cannot be grabbed from those central repos? The reason we have today AFAIK is for eclipse setup(In eclipse we set up a android project but not maven project, you have to put those jars somewhere, so eclipse will know their existence). But now we move to android studio, gradle build is integrated. Correct me if i'm wrong.

File: adal/build.gradle (hunk 2 above)
Comment (weijjia) [inline adal/build.gradle]: agree. :)

File: sample/build.gradle [hunk 10]
```diff
@@ -0,0 +1,33 @@
+apply plugin: 'com.android.application'
+
+android {
+    compileSdkVersion 21
+    buildToolsVersion "21.1.1"
```
Comment (weijjia) [inline sample/build.gradle]: why only sample app use build tool version 21.1.1 while others use 21.1.2?

File: sample/build.gradle [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    lintOptions {
+        abortOnError false
+    }
+}
+
+dependencies {
+    compile fileTree(dir: 'libs', include: ['*.jar'])
+    testCompile 'junit:junit:4.12'
```
Comment (weijjia) [inline sample/build.gradle]: Do we have tests for sample app? If so, why do we need tests for sample app?

File: adal/src/androidTest/java/com/microsoft/aad/adal/StorageHelperTests.java [hunk 12]
```diff
@@ -276,7 +279,7 @@ public void testKeyPair() throws
 
         assertTrue("Keystore has the alias", keyStore.containsAlias("AdalKey"));
     }
-
+*/
```
Comment (addev-ashish) [inline adal/src/androidTest/java/com/microsoft/aad/adal/StorageHelperTests.java]: is there a tab missing here? 
Also Why don't we just delete this function?

File: adal/build.gradle [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        }
+    }
+    sourceSets {
+        main {
+            manifest.srcFile 'src/main/AndroidManifest.xml'
+        }
+    }
+
+    lintOptions {
+        abortOnError false
```
Comment (agamemnum) [inline adal/build.gradle]: Is code fails Lint? Why do you disable this?

File: adal/build.gradle [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+        }
+    }
+
+}
+
+dependencies {
+    androidTestCompile 'org.mockito:mockito-core:1.10.19'
+    androidTestCompile 'com.google.dexmaker:dexmaker:1.2'
+    androidTestCompile 'com.google.dexmaker:dexmaker-mockito:1.2'
+    compile 'com.android.support:appcompat-v7:21.0.3'
```
Comment (agamemnum) [inline adal/build.gradle]: This appcompat is really old. Consider using at least 22

File: adal/build.gradle [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+            classpath 'com.github.dcendents:android-maven-gradle-plugin:1.3'
+    }
+}
+
+android {
+    compileSdkVersion 21
+    buildToolsVersion "21.1.2"
+
+    defaultConfig {
+        minSdkVersion 18
```
Comment (agamemnum) [inline adal/build.gradle]: it should be 14

File: adal/build.gradle (hunk 13 above)
Comment (heidijinxujia) [inline adal/build.gradle]: Hi @agamemnum , thank you for your review. The deployment of the gradle build is still under construction. The flag is set to be false temporarily, and I am fixing the lint errors these days. All the flags of code analysis tools will be set back to be true after finishing fixing all errors reported (ASAP). @RandalliLama

File: adal/build.gradle (hunk 14 above)
Comment (heidijinxujia) [inline adal/build.gradle]: OK. Thanks.

**Comment by addev-ashish** [comment]
#449 #448

**Comment by addev-ashish** [comment]
>                 return true;

This code should not change?

---
Refers to: adal/src/main/java/com/microsoft/aad/adal/AuthenticationActivity.java:625 in 0fb7c6f. [](commit_id = 0fb7c6faf6bcd68355db915a6748ef86387027e8, deletion_comment = False)

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationActivity.java [hunk 16]
```diff
@@ -959,13 +948,13 @@ private void setAccount(final TokenTaskResult result)
...
 
-            TokenCacheItem item = TokenCacheItem.createRegularTokenCacheItem(mRequest.getAuthority(), mRequest.getResource(), 
+            TokenCacheItem item = TokenCacheItem.createRegularTokenCacheItem(mRequest.getAuthority(), mRequest.getResource(),
                     mRequest.getClientId(), result.taskResult);
             String json = gson.toJson(item);
             String encrypted = mStorageHelper.encrypt(json);
 
             // Single user and cache is stored per account
-            String key = CacheKey.createCacheKeyForRTEntry(mAuthRequest.getAuthority(), mAuthRequest.getResource(), 
+            String key = CacheKey.createCacheKeyForRTEntry(mAuthRequest.getAuthority(), mAuthRequest.getResource(),
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationActivity.java]: looks like the FOCI change

**Pull request closed without merging**